File: Google/583.delete-operation-for-two-strings.java
/*
 * @lc app=leetcode id=583 lang=java
 *
 * [583] Delete Operation for Two Strings
 *
 * Only deletions are allowed, so the final equal string must be a common
 * subsequence. Keep the longest common subsequence; delete everything else.
 * A one-dimensional rolling DP computes LCS length.
 *
 * Time: O(|word1| * |word2|). Space: O(min lengths).
 */

// @lc code=start
class Solution {
    public int minDistance(String word1, String word2) {
        if (word2.length() > word1.length()) {
            String tmp = word1;
            word1 = word2;
            word2 = tmp;
        }

        int[] dp = new int[word2.length() + 1];
        for (int i = 0; i < word1.length(); i++) {
            int previousDiagonal = 0;
            for (int j = 1; j <= word2.length(); j++) {
                int previousRowSameColumn = dp[j];
                if (word1.charAt(i) == word2.charAt(j - 1)) {
                    dp[j] = previousDiagonal + 1;
                } else {
                    dp[j] = Math.max(dp[j], dp[j - 1]);
                }
                previousDiagonal = previousRowSameColumn;
            }
        }

        int lcs = dp[word2.length()];
        return word1.length() + word2.length() - 2 * lcs;
    }
}
// @lc code=end

File: companies/Apple/583.delete-operation-for-two-strings.go
package leetcode

//
// @lc app=leetcode id=583 lang=golang
//
// [583] Delete Operation for Two Strings
//
// Notes
// Minimum deletions equals len(word1)+len(word2)-2*LCS(word1,word2). Compute the
// LCS length with one rolling DP row over the shorter word. previousDiagonal
// stores dp[i-1][j-1] while dp[j] still holds the previous row's same-column
// value. Time: O(mn). Space: O(min(m,n)).
//
// @lc code=start

func MinDistance583(word1 string, word2 string) int {
	if len(word2) > len(word1) {
		word1, word2 = word2, word1
	}

	dp := make([]int, len(word2)+1)
	for i := range word1 {
		previousDiagonal := 0
		for index := 1; index <= len(word2); index++ {
			previousRowSameColumn := dp[index]
			if word1[i] == word2[index-1] {
				dp[index] = previousDiagonal + 1
			} else if dp[index-1] > dp[index] {
				dp[index] = dp[index-1]
			}
			previousDiagonal = previousRowSameColumn
		}
	}

	lcsLength := dp[len(word2)]
	return len(word1) + len(word2) - 2*lcsLength
}

// @lc code=end
